### DISCUSSION TITLE
.sorted method

### DISCUSSION THREAD

**Discussion by danieldaeschle** [vlang/v#11896] [Ideas]
like .sort it would be nice to have a .sorted method like in python.

`.x.sorted()` would (create a copy of the array) and return it. Using this technique, chaining methods is easier.

**Comment by dumblob** [comment]
I'd say such a method should **not** create a copy. For copy there is something like `.copy()` which could be then immediately followed by `.sorted()` as in-place sort and return the sorted array.

The reason I'm against `.sorted()` returning a copy is two-fold. First, it's faster :wink:. Second as noted above the *copy & sort* use case is easy and **explicit** but if `.sorted()` would internally make a copy, we couldn't satisfy the chaining requirement in case we wouldn't want to make a copy.

**Comment by danieldaeschle** [comment]
Imagine this case:
```
x := [1, 3, 2]
y := x.sorted()
```
If it wouldn't create a copy there are two problems: `x` is modified unexpectly, `x` needs to be `mut`. But for this case we have `.sort()`

**Comment by dumblob** [comment]
@danieldaeschle your example demonstrates what I've envisioned :wink:.

But I understand your concerns. How about leverage the fact `sort()` is already "magical" and have two `sorted()` methods with the same signature whereas one would dispatch on `mut` variable and would sort **in place** (i.e. without copying) and the second one would dispatch on non-`mut` variable and would **make a copy and then sort this copy** (both `sorted()` methods would return the resulting array for chaining).

Or alternatively change `sorted()` to something implying a copy - maybe `sortcopy()`?

**Comment by danieldaeschle** [comment]
Python has sort and sorted. I would vote for my initial request

**Comment by dumblob** [comment]
Or maybe stick with [*V* style](https://github.com/vlang/v/pull/10830/files#diff-e0f1f79ad24bb0fe6681f0cfc37ddc05ff9d4ba204b82a693ec0a2fe1db3def8R3823-R3827 ) (which seems to discourage chaining a bit) and not introduce `sorted()` at all?

**Comment by danieldaeschle** [comment]
I don't see the point of avoiding chaining methods. `filter`, `map` is suitable for chaining too.
If V tries to prevent nice coding styles it's better to avoid V for me.

`.sort` is really annoying because it needs a new line and breaks the chain. That's why I want the option of `.sorted`

**Comment by dumblob** [comment]
I understand your situation (from time to time I find myself writing simple wrappers over some of Pythons methods which do not return what I'd expect just to allow chaining).

It just seems to me it's not so black & white in case of *V* :wink: (starting from its `vfmt` which does at least two things I personally really dislike and going through the general style of using "overly many" variables everywhere to maintain highly vertical style).

There is also the option to simply return the sorted array from `.sort()` without introducing any new method (existing code would still work).

IDK what's best.

**Comment by JalonSolov** [comment]
Chaining can make nice, clean code, but it is almost impossible to debug without breaking the chain apart, assigning the intermediate values to temp variables, etc.

So why not just do that in the first place and avoid the problems with chaining?

**Comment by danieldaeschle** [comment]
>There is also the option to simply return the sorted array from .sort() without introducing any new method (existing code would still work).

I would be okay with that. But the variable needs to be `mut` I think and that's not very nice. The generated code is a C `qsort` statement which doesn't return anything.

>So why not just do that in the first place and avoid the problems with chaining?

If I need to debug it, I can split it into several lines. But I prefer a clean code style over a debuggable code style. AT LEAST THE POSSIBILITY COULD BE GIVEN.

**Comment by igotfr** [comment]
<s>in my opinion, the best name is `.clone_sort()` as the current way to do this is:
```v
a := [2,1,3]
mut b := a.clone()
b.sort()
```
</s>

**Comment by igotfr** [comment]
change the current `.sort()` method name  to `.sort_mutation()` or `.sort_mut()` and use the `.sort()` name to the method that not mutate the object would be a nicer, like in Ruby and Julia, but Ruby and Julia uses `!` to mutation methods

**Comment by danieldaeschle** [comment]
sort_mut is very verbose.

**Comment by igotfr** [comment]
> sort_mut is very verbose.

`sort_()`?

**Comment by JalonSolov** [comment]
`sort_()` is as "bad" as `sort!()`, which we decided not to use.

`sort()` is documented as mutating - that should be enough.

**Comment by danieldaeschle** [comment]
This is not about changing the sort() method but adding another method that does not mutate but returns a new array.

**Comment by igotfr** [comment]
> This is not about changing the sort() method but adding another method that does not mutate but returns a new array.

Python's sorted in the end, is the same as `.sort()` on Ruby and Julia, just Ruby and Julia have Naming convention and Python not

Python's sort is the same as `.sort!()` on Ruby and Julia

**Comment by igotfr** [comment]
> `sort_()` is as "bad" as `sort!()`, which we decided not to use.
> 
> `sort()` is documented as mutating - that should be enough.

I really don't understand why a Naming convention is bad

**Comment by igotfr** [comment]
https://github.com/vlang/v/issues/11439

**Comment by JalonSolov** [comment]
It can be "bad" for many reasons.  Adds complexity without any obvious meaning is one.  `!`, for example, usually means `not` in computers, so does `sort!()` mean `not sorted`?  (Well, `sorted not` the way it's written.)

Same for `sort_()`.  What does that even mean?  (without digging through the documentation...)

At least `sort_mutable()` is a bit better, though I'd turn it around to `mutable_sort()` I think.  But that's more verbose.

**Comment by igotfr** [comment]
the question is that V needs a Pattern for mutation methods

_ works as an implicit pattern like the keywords fn, pub, mut, etc

Others alternatives, but I prefer _

- _
- _mutation
- _mutable
- _mut
- _mt

sorted is bad, because the name is mutation method, not past method

**Comment by danieldaeschle** [comment]
My Idea was:

`sort` => mutates
`sorted` => does not mutate; creates a new array

**Comment by igotfr** [comment]
@danieldaeschle ok, but `sorted` is a bad name

**Comment by danieldaeschle** [comment]
Explain further...?

**Comment by igotfr** [comment]
@danieldaeschle I suggested a convention Naming on mutation methods because is the pattern used in Ruby and Julia, but it's okay a convention Naming on non-mutation methods. However, sorted is past verb, a better name needs relate to non-mutation methods (I don't have suggestions for this)

maybe:
`_`
`_clone`
